### PR TITLE
[cleanup] multiprocess_pipe: dead-code removal and simplification

### PR DESCRIPTION
**Pull request by msbaines**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/pipe/multiprocess_pipe.py [hunk 1]
```diff
@@ -463,7 +461,6 @@ def __init__(
 
         self.group = group
         self.worker_map = worker_map
-        self.input_device = input_device
```
Comment (sidgoyal78) [inline fairscale/nn/pipe/multiprocess_pipe.py:466]: We are using this in ampnet
https://github.com/facebookresearch/fairscale/blob/master/experimental/nn/ampnet_pipe/pipe.py#L60

File: fairscale/nn/pipe/async_schedule.py [hunk 2]
```diff
@@ -194,13 +194,7 @@ def run_invocation(
...
 
         task = create_task(
-            PipelineStyle.AsyncSchedule,
-            self.checkpoint_stop,
-            batch.index,
-            self.group.rank(),
-            batch,
-            partition.module,
-            skip_trackers,
+            self.checkpoint_stop, batch.index, self.group.rank(), batch, partition.module, skip_trackers,
```
Comment (sidgoyal78) [inline fairscale/nn/pipe/async_schedule.py:197]: I think this is needed in the multiprocess_pipeline.py function in the run() function to distinguish the even-loop and non-event-loop cases.

**Review by sidgoyal78 — CHANGES_REQUESTED** [review]
Thanks for the PR, Mandeep. I have put a few inline comments:

**Review by min-xu-ai — APPROVED** [review]
nice. looks shorter and cleaner

File: fairscale/nn/pipe/async_schedule.py [hunk 3]
```diff
@@ -194,13 +194,7 @@ def run_invocation(
         from .multiprocess_pipeline import create_task
```
Comment (anj-s) [inline fairscale/nn/pipe/async_schedule.py:194]: can we move this to the top of the file?

File: fairscale/nn/pipe/multiprocess_pipe.py [hunk 4]
```diff
@@ -622,7 +613,7 @@ def back_helper(self, output: List[microbatch.Batch]) -> None:
             raise ValueError("back_helper should only be called on non-final stages")
 
         if self.pipeline:
-            self.pipeline.back_helper(list(reversed(output)))
+            self.pipeline.back_helper(output)
```
[the remark this answers is not in the document]
Comment (msbaines) [reply fairscale/nn/pipe/multiprocess_pipe.py:617]: I am moving the reversed from here to multiprocess_pipeline. The code feels clearer this way. I prefer passing the unmodified output and then reversing when I need to. It was not obvious in multiprocess_pipeline that the output was reversed.

File: fairscale/nn/pipe/async_schedule.py (hunk 2 above)
Comment (msbaines) [reply fairscale/nn/pipe/async_schedule.py:197]: Yes. But it is not used inside create_task.

File: fairscale/nn/pipe/multiprocess_pipe.py (hunk 1 above)
Comment (msbaines) [reply fairscale/nn/pipe/multiprocess_pipe.py]: Ah. Good catch. Fixed.

File: fairscale/nn/pipe/multiprocess_pipeline.py [hunk 5]
```diff
@@ -325,7 +319,6 @@ def compute(
         """Runs tasks with synchronization to copy streams."""
```
Comment (anj-s) [inline fairscale/nn/pipe/multiprocess_pipeline.py:320]: If we do a quick return here for any style that is not MultiProcess, can we get rid of the two if statements below?

**Comment by anj-s** [comment]
Would it be possible to add more comments to the code as you do the cleanup? It would really help understand this piece of code better. Thanks!

File: fairscale/nn/pipe/multiprocess_pipeline.py (hunk 5 above)
Comment (msbaines) [reply fairscale/nn/pipe/multiprocess_pipeline.py:320]: Great catch! Fixed.

File: fairscale/nn/pipe/async_schedule.py (hunk 3 above)
Comment (msbaines) [reply fairscale/nn/pipe/async_schedule.py:196]: This is needed here to avoid a cyclic dependency. Added a TODO to remove the cycle.

**Review by sidgoyal78 — APPROVED** [review]
Thanks for the changes. Looks good.

**Pull request merged**